src/org/omegat/util/StringUtil.java

```diff
@@ -511,4 +511,31 @@ public static String format(String str, Object... arguments) {
         str = str.replaceAll("'", "''");
         return MessageFormat.format(str, arguments);
     }
+
+    /**
+     * convert Zenkaku/Hankaku alphabet to regex  that match both
+     *
+     * replace (\uFF01 - \uFF5E) to (\u0021- \u007E)
+     * such as text = text.replaceAll("u\0021", "(\uFF01|\u0021)");
+     * except for chars that consist of regex.
+     * FIXME: handling 0x20-0x2f, 0x5c
+     *
+     * @param text
+     * @return String with regex
+     */
+    public static String replaceZenHan(String text) {
```

Comment:
Why do you think steps 2–5 are necessary? We only need to normalize the texts to one form (half-width or full-width) and then do a normal comparison. There is no need to build a regex.